### PR TITLE
Stable Diffusion CUDA Optimizations

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description

Add stable diffusion CUDA kernel optimizations.

The following are included:
(1) GroupNorm operator. This kernel is from TensorRT 8.5.
(2) BiasSplitGelu operator. This kernel is modified from SplitGelu of TensorRT 8.5. We added bias to the SplitGelu.
(3) NhwcConv operator. This adds support of NHWC format (ONNX Conv operator uses NCHW format).
(3) Update MultiHeadAttention (packed kv and no bias) for cross attention. This could avoid transpose of kv for TRT fused cross attention kernel.
(4) Optimization and benchmark script

Not included:
(1) Script to convert Conv to NhwcConv in onnx graph.
(2) Update symbolic shape inference for NhwcConv.
(3) Add SeqLen2Spatial operator
(4) Documents

Limitations: GroupNorm, BiasSplitGelu and NhwcConv kernels are implemented based on stable diffusion usage. They might not be applicable to any input size or dimensions. For example, BiasSplitGelu requires hidden size to be 2560 | 5120 | 10240, and NhwcConv assumes 4D input/weight.

There is minor increasement of binary size. For SM=75 only, python package wheel size adds (33757K - 33640K) = 117 KB. It is possible to move NHWC from template parameter to constructor to reduce binary size (with slight cost of performance).

Note: for RTX 4090/4080/4070 Ti, need build with CUDA 11.8 and latest cuDNN to get best performance.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/diffusion/group_norm.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#include "core/common/common.h"
+#include "core/providers/cuda/cuda_kernel.h"
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+using namespace onnxruntime::cuda;
+
+template <typename T>
```
Comment (yuslepukhin) [inline onnxruntime/contrib_ops/cuda/diffusion/group_norm.h:14]: >emplate <typename T> [](http://example.com/codeflow?start=1&length=20)

Can we make only ComputeInternal() templated? The constructor does not need a template, does it?

File: onnxruntime/contrib_ops/cuda/diffusion/group_norm.cc [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  int height = static_cast<int>(input_dims[2]);
+  int width = static_cast<int>(input_dims[3]);
+
+  if (num_channels % num_groups_ != 0) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                            "number of channels should be divisiable by num_groups");
+  }
+
+  auto workspace = GetScratchBuffer<void>(GetGroupNormWorkspaceSizeInBytes(), context->GetComputeStream());
+
```
Comment (yuslepukhin) [inline onnxruntime/contrib_ops/cuda/diffusion/group_norm.cc:85]: Most of this code is not type dependent as well as the constructor above. Can we move it to a non-templated helpers?

File: onnxruntime/contrib_ops/cuda/diffusion/group_norm_impl.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+constexpr size_t kMaxGroupNormBatchSize = 32;
+constexpr size_t kGroupNormNumberOfGroups = 32;
+
+constexpr size_t GetGroupNormWorkspaceSizeInBytes()
+{
```
Comment (yuslepukhin) [inline onnxruntime/contrib_ops/cuda/diffusion/group_norm_impl.h:19]: nit: opening bracket. Use .clangformat

File: onnxruntime/contrib_ops/cuda/diffusion/group_norm.cc (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/diffusion/group_norm.cc:115]: updated class to no template.

File: onnxruntime/contrib_ops/cuda/diffusion/group_norm.h (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/diffusion/group_norm.h]: updated class to no template.

File: onnxruntime/contrib_ops/cuda/diffusion/bias_split_gelu.cc [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "bias is expected to have 1 dimensions, got ", bias_dims.size());
+  }
+  if (bias_dims[0] != input_dims[2]) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "last dimension of input and bias are not the same");
+  }
+
+  TensorShapeVector output_shape = input->Shape().AsShapeVector();
+  output_shape[2] = input_dims[2] / 2;
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/diffusion/bias_split_gelu.cc:54]: need it ensure input_dims[2] is divided by 2?

Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/diffusion/bias_split_gelu.cc:54]: BiasSplitGelu requires hidden size to be 2560 | 5120 | 10240. 

It's fine to not add a check. The worst case is user intentionally uses hidden size 2561 and still get result (odd number cannot be split evenly so expect output is not defined).

**Review by wangyems — DISMISSED** [review]
lgtm

File: onnxruntime/contrib_ops/cuda/diffusion/bias_split_gelu.cc (hunk 4 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/diffusion/bias_split_gelu.cc:59]: updated: add a check of hidden_size be 2560 | 5120 | 10240.

**Review by yufenglee — APPROVED** [review]

**Review by wangyems — APPROVED** [review]

**Pull request merged**